### PR TITLE
Nullable ref type annotation fixes to analyzer APIs

### PR DESCRIPTION
**Pull request by AArnott**
The first commit is pretty straightforward as correct.
My second commit might be a bit more controversial, so I kept it separate in case we want to remove it from the branch before merging.

### CODE REVIEW

File: src/Compilers/Core/Portable/CommandLine/SarifErrorLogger.cs [hunk 1]
```diff
@@ -103,7 +103,7 @@ protected void WriteResultProperties(Diagnostic diagnostic)
 
                     foreach (var pair in diagnostic.Properties.OrderBy(x => x.Key, StringComparer.Ordinal))
                     {
-                        _writer.Write(pair.Key, pair.Value);
+                        _writer.Write(pair.Key, pair.Value ?? string.Empty);
```
Comment (AArnott) [inline src/Compilers/Core/Portable/CommandLine/SarifErrorLogger.cs:106]: Before my change (and before/ignoring annotations), a null value gets serialized as if it's the empty string.
My change here avoids a compiler warning due to nullable annotations.

Comment (cston) [reply src/Compilers/Core/Portable/CommandLine/SarifErrorLogger.cs:106]: It looks like `JsonWriter.Write(string, string)` should be annotated as `Write(string, string?)` instead.

---
In reply to: [402565899](https://github.com/dotnet/roslyn/pull/43023#discussion_r402565899) [](ancestors = 402565899)

**Comment by sharwell** [comment]
Why are we allowing messageargs to be null?

**Comment by AArnott** [comment]
> Why are we allowing messageargs to be null?

Roslyn always allowed this (and in some code paths explicitly pass `null` in to this `params` argument. But the annotations suggest it's not allowed. I'm fixing the annotations to match the actual supported behavior.

File: src/Compilers/Core/Portable/CommandLine/SarifErrorLogger.cs (hunk 1 above)
Comment (cston) [reply src/Compilers/Core/Portable/CommandLine/SarifErrorLogger.cs:106]: > However I think it's misleading because JsonWriter.Write(string, string) does not preserve null values. It converts them to "". 

The annotation would indicate the method accepts a `null` value, it would not indicate how the value is used. The annotation on the `location` parameter in `Diagnostic.Create()` is similar. In that case, a `null` location is converted to `Location.None`.

---
In reply to: [402592535](https://github.com/dotnet/roslyn/pull/43023#discussion_r402592535) [](ancestors = 402592535)

Comment (AArnott) [reply src/Compilers/Core/Portable/CommandLine/SarifErrorLogger.cs:106]: I've pushed a change to add that annotation.

Comment (cston) [inline src/Compilers/Core/Portable/CommandLine/SarifErrorLogger.cs:106]: > ?? string.Empty [](start = 58, length = 16)

Is this change still needed?

Comment (AArnott) [reply src/Compilers/Core/Portable/CommandLine/SarifErrorLogger.cs]: Nope. I've reverted this change.

**Comment by cston** [comment]
@AArnott, it looks like some uses of `Diagnostic.Properties` are now resulting in nullability warnings.

**Comment by AArnott** [comment]
I was only checking against Compilers.sln I see in Roslyn.sln there are some warnings. I'll look into it, @cston .

File: src/Analyzers/Core/CodeFixes/PopulateSwitch/AbstractPopulateSwitchCodeFixProvider.cs [hunk 2]
```diff
@@ -58,8 +58,8 @@ public sealed override Task RegisterCodeFixesAsync(CodeFixContext context)
         {
             var diagnostic = context.Diagnostics.First();
             var properties = diagnostic.Properties;
-            var missingCases = bool.Parse(properties[PopulateSwitchStatementHelpers.MissingCases]);
-            var missingDefaultCase = bool.Parse(properties[PopulateSwitchStatementHelpers.MissingDefaultCase]);
+            var missingCases = bool.Parse(properties[PopulateSwitchStatementHelpers.MissingCases] ?? string.Empty);
```
Comment (cston) [inline src/Analyzers/Core/CodeFixes/PopulateSwitch/AbstractPopulateSwitchCodeFixProvider.cs:61]: >string.Empty [](start = 101, length = 12)

`bool.Parse()` will throw an exception parsing `string.Empty` (or `null`). Presumably we're not passing those values now so perhaps we should use `bool.Parse(properties[...]!)` instead.

File: src/Analyzers/Core/CodeFixes/PopulateSwitch/AbstractPopulateSwitchCodeFixProvider.cs [hunk 3]
```diff
@@ -137,8 +137,8 @@ public sealed override Task RegisterCodeFixesAsync(CodeFixContext context)
             bool addCases, bool addDefaultCase, bool onlyOneDiagnostic,
             CancellationToken cancellationToken)
         {
-            var hasMissingCases = bool.Parse(diagnostic.Properties[PopulateSwitchStatementHelpers.MissingCases]);
-            var hasMissingDefaultCase = bool.Parse(diagnostic.Properties[PopulateSwitchStatementHelpers.MissingDefaultCase]);
+            var hasMissingCases = bool.Parse(diagnostic.Properties[PopulateSwitchStatementHelpers.MissingCases] ?? string.Empty);
```
Comment (cston) [inline src/Analyzers/Core/CodeFixes/PopulateSwitch/AbstractPopulateSwitchCodeFixProvider.cs:140]: > ?? string.Empty [](start = 111, length = 16)

`!` here and below.

File: src/Analyzers/Core/CodeFixes/PopulateSwitch/AbstractPopulateSwitchCodeFixProvider.cs (hunk 2 above)
Comment (AArnott) [reply src/Analyzers/Core/CodeFixes/PopulateSwitch/AbstractPopulateSwitchCodeFixProvider.cs:61]: True, `""` will throw, just like `null` would (but different exceptions, and with this change we don't get a build warning). 

Sure, I could use `!` but there's in fact no guarantee that the value *isn't* null as far as I could tell. I suppose if the code fix provider wants to make an assumption about the analyzer never setting a null value, that's fine. I'll change to the `!` here.

File: src/Analyzers/Core/CodeFixes/PopulateSwitch/AbstractPopulateSwitchCodeFixProvider.cs (hunk 3 above)
Comment (AArnott) [reply src/Analyzers/Core/CodeFixes/PopulateSwitch/AbstractPopulateSwitchCodeFixProvider.cs:140]: Done.

**Comment by AArnott** [comment]
Any other reviewers before this gets merged?

File: src/Analyzers/Core/Analyzers/PopulateSwitch/AbstractPopulateSwitchDiagnosticAnalyzer.cs [hunk 4]
```diff
@@ -53,7 +53,7 @@ private void AnalyzeOperation(OperationAnalysisContext context)
                 !tree.OverlapsHiddenPosition(switchBlock.Span, context.CancellationToken))
             {
                 Debug.Assert(missingCases || missingDefaultCase);
-                var properties = ImmutableDictionary<string, string>.Empty
+                var properties = ImmutableDictionary<string, string?>.Empty
```
Comment (CyrusNajmabadi) [inline src/Analyzers/Core/Analyzers/PopulateSwitch/AbstractPopulateSwitchDiagnosticAnalyzer.cs:56]: hrmm... who passes null for the value.  I'd prefer we attempt to disallow that.  If you make this non-null, can you point to any violators?

Comment (mavasani) [reply src/Analyzers/Core/Analyzers/PopulateSwitch/AbstractPopulateSwitchDiagnosticAnalyzer.cs:56]: I don't think we can restrict these the values for the property bag to be non-null. I think I have seen some third party analyzers using null as a value for the property bag entry.

File: src/Analyzers/Core/CodeFixes/PopulateSwitch/AbstractPopulateSwitchCodeFixProvider.cs [hunk 5]
```diff
@@ -58,8 +58,8 @@ public sealed override Task RegisterCodeFixesAsync(CodeFixContext context)
         {
             var diagnostic = context.Diagnostics.First();
             var properties = diagnostic.Properties;
-            var missingCases = bool.Parse(properties[PopulateSwitchStatementHelpers.MissingCases]);
-            var missingDefaultCase = bool.Parse(properties[PopulateSwitchStatementHelpers.MissingDefaultCase]);
+            var missingCases = bool.Parse(properties[PopulateSwitchStatementHelpers.MissingCases]!);
```
Comment (jcouv) [inline src/Analyzers/Core/CodeFixes/PopulateSwitch/AbstractPopulateSwitchCodeFixProvider.cs:61]: >! [](start = 97, length = 1)

We typically reserve the `!` suppressions for things that are checked in the current method (directly or via straightforward helper method) but the compiler is unable to track.
For more complicated invariants, we use assertions instead (`Debug.Assert(... is object);` or `is { }` depending on preference).
These two (and two more below in this file) strike me as cases for assertions.

[https://github.com/dotnet/roslyn/blob/master/docs/contributing/Nullable%20Annotations.md](https://github.com/dotnet/roslyn/blob/master/docs/contributing/Nullable%20Annotations.md)

Comment (cston) [reply src/Analyzers/Core/CodeFixes/PopulateSwitch/AbstractPopulateSwitchCodeFixProvider.cs:61]: An assert seems unnecessary here since `bool.Parse()` will throw an exception if the argument is `null`.

---
In reply to: [405655126](https://github.com/dotnet/roslyn/pull/43023#discussion_r405655126) [](ancestors = 405655126)

Comment (AArnott) [reply src/Analyzers/Core/CodeFixes/PopulateSwitch/AbstractPopulateSwitchCodeFixProvider.cs:61]: Ya. The ! was just to keep the compiler warning quiet so that the original behavior of throwing from bool.Parse would be preserved.

Comment (jcouv) [reply src/Analyzers/Core/CodeFixes/PopulateSwitch/AbstractPopulateSwitchCodeFixProvider.cs:61]: Makes sense. Thanks
